File: lib/nodejs/src/CmsClient.js
import thrift from 'thrift';
import url from 'url';
import AdminMenuService from '../lib/thrift/AdminMenuService';
import AdminTagService from '../lib/thrift/AdminTagService';
import AdminUserService from '../lib/thrift/AdminUserService';
import ClientWrapper from './ClientWrapper';

class CmsClient {
  constructor(endpoint = 'http://localhost') {
    const urlObj = url.parse(endpoint);
    const connection = thrift.createHttpConnection(urlObj.hostname, urlObj.port, {
      path: urlObj.pathname,
      https: urlObj.protocol === 'https:',
      transport: thrift.TBufferedTransport,
      protocol: thrift.TJSONProtocol,
      headers: {
        Accept: 'application/x-thrift',
      },
    });

    this.adminMenu = new ClientWrapper('AdminMenu', AdminMenuService, connection);
    this.adminTag = new ClientWrapper('AdminTag', AdminTagService, connection);
    this.adminUser = new ClientWrapper('AdminUser', AdminUserService, connection);
  }
}

export default CmsClient;

File: lib/python/src/CmsClient.py
# -*- coding: utf-8 -*-

from thrift import Thrift
from thrift.transport import THttpClient
from thrift.protocol import TMultiplexedProtocol
from thrift.protocol import TJSONProtocol
from AdminMenu import AdminMenuService
from AdminTag import AdminTagService
from AdminUser import AdminUserService

class CmsClient:
    def __init__(self, uriOrHost, port=None, path=None):
        self.uriOrHost = uriOrHost
        self.port = port
        self.path = path
        self.adminMenu = self.createClient('AdminMenu', AdminMenuService)
        self.adminTag = self.createClient('AdminTag', AdminTagService)
        self.adminUser = self.createClient('AdminUser', AdminUserService)

    def createClient(self, serviceName, service):
        client = THttpClient.THttpClient(self.uriOrHost, self.port, self.path)
        client.setCustomHeaders({'Accept' : 'application/x-thrift'})
        protocol = TJSONProtocol.TJSONProtocol(client)
        protocol = TMultiplexedProtocol.TMultiplexedProtocol(protocol, serviceName)
        return service.Client(protocol)
